File: src/quran/native/MushafLineView.ts
/**
 * The native line view: a Ḥafṣ line drawn straight onto a canvas.
 *
 * ── WHY NOT <Text> ────────────────────────────────────────────────────
 *
 * A line of the muṣḥaf is one run of QPC glyphs, one glyph per word, laid
 * out at the font's own advances plus a gap the layout solved for. `<Text>`
 * can draw that, and did, at two costs that are the platform's, not ours:
 *
 *   • it BREAKS a line that comes out a hair wider than its box — between
 *     glyphs when nothing else will do, and a glyph is a whole word — so
 *     every line's box carried half an em of slack it never used, and the
 *     page was set that much smaller (`MUSHAF_LINE_BOX_SLACK_EM`);
 *   • it CLIPS the ink to the view, so calligraphy that overshoots its
 *     metrics — every swash, every stacked maddah — needs the view padded
 *     out and the padding pulled back with margins (`lineInkPadding`).
 *
 * The native view does neither. It measures each word with the same
 * shaper the platform's text view uses (Minikin on Android, CoreText on
 * iOS and Catalyst), puts the pen exactly where the layout says, and
 * draws: no box to overflow, no last word to lose. The ink still needs
 * room to land in, and the view is sized for it, but the line's BOX is the
 * run itself, so the widest line spans the whole text block.
 *
 * The `<Text>` path stays as the fallback — under Jest, on a build whose
 * native side lacks the view, and for one release as the escape hatch —
 * and everything else about a page (hit-testing, the word reader, the
 * follow-scroll, the scrubber's thumbnails) reads the block width from
 * `pageBlockEm`, which knows which path is drawing.
 */
import type { HostComponent, Platform, UIManager } from 'react-native';
import type { NativeProps, NativeRun } from './MushafLineNativeComponent';

export const MUSHAF_LINE_VIEW = 'MushafLine';

/**
 * One piece of a line, in drawing order (right to left): a run of glyphs
 * (`t`, in the page font), or a gap the pen skips (`g`, its width in dp).
 * A wash (`w`) is drawn behind the piece; an ink colour (`i`) replaces the
 * line's for the glyphs — only ever the marker's medallion. Both as
 * `processColor` gives them.
 *
 * The view's other props: `fontFamily` and `fontSize` (dp); `color`, the
 * page's ink; `penRight`, the x of the run's RIGHT edge inside the view
 * (the text is RTL); `penBaseline`, its y; and `boxTop` / `boxHeight`,
 * the line box the washes fill. All in dp from the view's own top-left.
 */
export type MushafLineRun = NativeRun;
export type MushafLineNativeProps = NativeProps;

let available: boolean | null = null;

/**
 * Whether this build carries the view. Decided once, on first use rather
 * than at import — the UI manager is asked, and it answers for a legacy
 * view manager on the new architecture the same way it does for a Fabric
 * component. Under Jest the answer is no, and the `<Text>` path draws.
 *
 * React Native is required here, not imported above: `mushafLayout.ts`
 * asks this to size the page block, and that module stays free of React
 * Native so the tests can replay exactly what the renderer draws.
 */
export function nativeMushafLineAvailable(): boolean {
  if (available === null) {
    try {
      const rn = require('react-native') as {
        Platform: typeof Platform;
        UIManager: typeof UIManager;
      };
      available =
        rn.Platform.OS !== 'web' &&
        typeof rn.UIManager.hasViewManagerConfig === 'function' &&
        rn.UIManager.hasViewManagerConfig(MUSHAF_LINE_VIEW) === true;
    } catch {
      available = false;
    }
  }
  return available;
}

/** Tests only: pretend the view is (or is not) there. */
export function setNativeMushafLineAvailableForTests(value: boolean | null): void {
  available = value;
}

let component: HostComponent<MushafLineNativeProps> | null = null;

/**
 * The host component — call only after `nativeMushafLineAvailable()`.
 * Required lazily so that a build without the view never registers it.
 */
export function MushafLineNative(): HostComponent<MushafLineNativeProps> {
  if (component === null) {
    component = (
      require('./MushafLineNativeComponent') as {
        default: HostComponent<MushafLineNativeProps>;
      }
    ).default;
  }
  return component;
}

File: android/app/src/main/java/com/prayer_times/MushafLineView.kt
package com.prayer_times

import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Typeface
import android.view.View
import com.facebook.react.bridge.ReadableArray
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.common.assets.ReactFontManager

/**
 * One line of the Ḥafṣ muṣḥaf, drawn with the pen.
 *
 * The line arrives as pieces in drawing order — right to left, the text is
 * RTL — each a run of glyphs in the page font or a gap the pen skips, with
 * an optional wash behind it and, for the marker's medallion, its own ink.
 * The pen starts at `penRight` and walks left by each run's measured advance
 * (Minikin's, the same shaper the platform's TextView uses, so a word is
 * exactly as wide here as it was there) or by the gap's width.
 *
 * Nothing here breaks or clips: a run a hair wider than expected lands a
 * hair further left, and the ink that overshoots the line's ends and its
 * metrics lands wherever the view has room for it — the view is sized by
 * JS to have that room (`lineInkPadding`, `lineInkSidePadding`). See
 * `src/quran/native/MushafLineView.ts` for why this exists.
 *
 * All positions are in dp from JS and scaled here.
 */
class MushafLineView(context: Context) : View(context) {
  var fontFamily: String = ""
    set(value) { field = value; invalidate() }
  var fontSize: Float = 0f
    set(value) { field = value; invalidate() }
  var color: Int = -0x1000000
    set(value) { field = value; invalidate() }
  var runs: ReadableArray? = null
    set(value) { field = value; invalidate() }
  var penRight: Float = 0f
    set(value) { field = value; invalidate() }
  var penBaseline: Float = 0f
    set(value) { field = value; invalidate() }
  var boxTop: Float = 0f
    set(value) { field = value; invalidate() }
  var boxHeight: Float = 0f
    set(value) { field = value; invalidate() }

  private val glyphPaint = Paint(Paint.ANTI_ALIAS_FLAG or Paint.SUBPIXEL_TEXT_FLAG).apply {
    textAlign = Paint.Align.LEFT
  }
  private val washPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply { style = Paint.Style.FILL }

  init {
    // The view is transparent and never needs a hardware layer of its own.
    setWillNotDraw(false)
  }

  override fun onDraw(canvas: Canvas) {
    super.onDraw(canvas)
    val pieces = runs ?: return
    if (fontSize <= 0f || fontFamily.isEmpty() || pieces.size() == 0) return
    val density = resources.displayMetrics.density
    // Resolved at draw time, not when the prop lands: the font store can
    // re-register a slot family with another page's file, and a page that
    // is on screen is pinned against that — but a lookup is a map read,
    // and it is the one thing that can never go stale.
    glyphPaint.typeface =
      ReactFontManager.getInstance().getTypeface(fontFamily, Typeface.NORMAL, context.assets)
    glyphPaint.textSize = fontSize * density
    glyphPaint.letterSpacing = 0f

    val top = boxTop * density
    val bottom = (boxTop + boxHeight) * density
    val y = penBaseline * density
    val count = pieces.size()

    // Two passes, washes then glyphs, so a swash that crosses into the
    // next word is drawn over its wash and not under it — which is how a
    // paragraph draws its span backgrounds too.
    var pen = penRight * density
    val lefts = FloatArray(count)
    val texts = arrayOfNulls<String>(count)
    for (i in 0 until count) {
      val piece: ReadableMap = pieces.getMap(i) ?: continue
      val advance: Float
      if (piece.hasKey("t") && !piece.isNull("t")) {
        val text = piece.getString("t") ?: ""
        texts[i] = text
        advance = glyphPaint.measureText(text)
      } else {
        advance = (if (piece.hasKey("g")) piece.getDouble("g").toFloat() else 0f) * density
      }
      val left = pen - advance
      lefts[i] = left
      if (piece.hasKey("w") && !piece.isNull("w")) {
        washPaint.color = piece.getInt("w")
        canvas.drawRect(left, top, pen, bottom, washPaint)
      }
      pen = left
    }
    for (i in 0 until count) {
      val text = texts[i] ?: continue
      val piece = pieces.getMap(i) ?: continue
      glyphPaint.color =
        if (piece.hasKey("i") && !piece.isNull("i")) piece.getInt("i") else color
      canvas.drawText(text, lefts[i], y, glyphPaint)
    }
  }
}
